### ISSUE TITLE
Cross-stock attention block has no residual path and discards 94% of the cross-section

### ISSUE THREAD

**Issue by magilliam27** [magilliam27/MCI-GRU#197]
**Mode:** AFK. Fully specified; an agent can resolve this alone.

**Owned paths:**

- `mci_gru/models/attention.py`
- `mci_gru/models/trunk.py`
- `mci_gru/models/factory.py`
- `mci_gru/config.py`
- `tests/test_cross_section_block.py` (new)
- `docs/CONFIGURATION_GUIDE.md` (one row in the model table)

No other path is touched. In particular **no default changes**: `configs/config.yaml`, every preset under `configs/experiment/`, and the frozen recipe in `docs/DEFAULT_EXPERIMENT_RECIPE.md` are all out of scope.

## Summary

`SelfAttention`, the cross-stock block in the trunk, is applied once and its output **replaces** `z`. It has no residual path, no normalisation, and no feed-forward sublayer:

```python
# mci_gru/models/trunk.py, StockPredictionModel.forward
z = z.view(batch_size, num_stocks, -1)
z = self.self_attention(z, stock_mask=stock_mask)   # replaces, does not correct
z = z.view(batch_size * num_stocks, -1)
```

This is the rank-collapse failure that Dong et al. (2021) prove for attention without skip connections. With a single 128-wide head, the attention logits start near-uniform, so every stock receives approximately the same average of all stocks, and whatever distinguished a stock is discarded before the score head sees it.

## Evidence

Measured on the real universe (`configs/data/gics_top10_110_2016.yaml`: 201-node PIT union axis, ~110 admissible names per session, masked-panel mode, 2,012 training days), on all 239 validation dates, with the PIT mask applied as the trainer applies it, on a **trained** checkpoint. Full method and raw output in `docs/research/current/MCI_GRU_TRUNK_ARCHITECTURE_OPPORTUNITIES_2026-09-05.md` section 3.7 and the `trunk_architecture_diagnostics_2026-09-05/` folder beside it.

| Quantity, trained model | Shipped block | With a residual path |
| --- | --- | --- |
| Cross-sectional variance share, before to after the block | **0.433 to 0.024** | 0.403 to 0.329 |
| Effective rank, before to after | 3.6 to 1.6 | 3.5 to 3.6 |
| Attention entropy (uniform bound 5.30) | 5.280 | 5.281 |
| Mean self-weight (uniform 0.0050) | 0.0048 | 0.0046 |

Three things this says:

1. The block discards **94 percent** of the variation between stocks on a date.
2. After training it is still uniform to within 0.02 nats, and it weights a stock on itself *below* uniform. It is not performing selective attention between stocks; it is passing a date-level average.
3. The block holds **49,664 of the model's 86,762 parameters**, 56 percent, to compute something the ranking objective largely cannot use, because `ICLoss` centres predictions within each date and a common shift is invisible to it.

## What this issue does NOT claim

**There is no evidence that fixing this improves IC, and it should not be represented as a performance fix.** Two-seed smokes on the real universe put the residual variant last on one seed and first on the other, with a within-variant seed spread (0.0257 test IC) larger than every between-variant gap and roughly seven times the paired protocol's 0.0035 detectable effect. See section 3.8 of the same document.

This is filed as a **correctness defect**: the block does not do the thing its name and position say it does, it is not the standard construction, and it is unstable at higher learning rates (section 3.2.2). The fix is 256 parameters and no measurable step-time cost. Whether it earns return is for the ablation in section 5 of the map to decide, not for this issue.

## Seams under test

Per `/mattpocock-skills:tdd`, the seams are declared here before any test is written.

**Tested at:**

- `mci_gru.models.create_model(input_size, config) -> StockPredictionModel` — construction and checkpoint compatibility.
- `StockPredictionModel.forward(...)` — end-to-end output shape, masking, and the anti-collapse behaviour.
- The cross-stock block's own `forward(x, stock_mask=None)` — the module boundary `tests/test_pit_masked_panel.py::test_self_attention_mask_prevents_inactive_node_influence` already uses.
- `ModelConfig` construction and `ModelConfig.to_dict()` — validation and serialisation.

**Explicitly not tested:** attention weight values, entropy or logit statistics, effective rank, or any private helper. Those are diagnostics that belong in the research map, not assertions in the suite; asserting on them would couple the tests to the implementation.

## Proposed change

Add a config-gated residual pre-norm form, default off.

1. `ResidualCrossSectionBlock` in `mci_gru/models/attention.py`, wrapping the existing `SelfAttention`:

   ```python
   out = x + self.inner(self.ln(x), stock_mask=stock_mask)
   if stock_mask is not None:
       out = out * stock_mask.unsqueeze(-1).to(out.dtype)
   return out
   ```

2. `ModelConfig.cross_section_block: str = "legacy"`, valid `{"legacy", "residual"}`, validated in `__post_init__` and serialised by `to_dict`.
3. Routed through `create_model` with a legacy-safe `config.get(...)` default so `config.yaml` files inside existing checkpoint directories build the identical module graph.
4. The `[A1, A2, B1, B2]` stream order and the group-type embedding contract are untouched; the block still consumes the same `z`.

Multi-head splitting and an optional feed-forward sublayer are **out of scope** for this issue. They are separate variables and would confound the arm.

## Acceptance criteria

- [ ] With default config, `create_model` produces a model whose `state_dict()` keys and parameter count are unchanged, and a state dict saved before this change loads with `strict=True`.
- [ ] With `cross_section_block="residual"`, forward returns `(batch, stocks)` and gradients reach every parameter of the block.
- [ ] Distinct per-stock inputs do not produce identical per-stock representations through the block.
- [ ] PIT masking holds: inactive nodes are exactly zero at the output, and perturbing an inactive stock does not move any active stock's score.
- [ ] Finite output under `torch.autocast` on CPU bfloat16.
- [ ] `ModelConfig` rejects an invalid value and `to_dict()` round-trips the new field.
- [ ] Every load-bearing test mutation-checked, with the table posted on the pull request, testing **inversions as well as deletions**.
- [ ] `ruff check .` and `ruff format --check .` clean; full suite green via `scripts/run_pytest_isolated.py`.
- [ ] `docs/TEST_REGISTRY.md` regenerated.

## Out of scope

- Any default, preset, or frozen-recipe change.
- Multi-head attention or an FFN sublayer in the block.
- The `MarketLatentStateLearner` defect, which is filed separately.
- Deleting the block (`use_self_attention=false`); that flag already exists and is an ablation arm, not a code change.
- Any performance claim.

**Comment by magilliam27** [comment]
**Owned paths expanded during implementation.** Recording it here rather than only in the pull request, since a concurrent session reads the ticket and not the branch.

Added to the declared list:

| Path | Why |
| --- | --- |
| `configs/config.yaml` | The two-axis review flagged a hard breach of `AGENTS.md` ("Config changes go through Hydra YAML in `configs/`"). Without a row here the knob needs `+model.cross_section_block=residual` append syntax rather than plain override, which would bite the section 5 ablation. Added at value `legacy`, which **is** the existing default, so this is not a default change. |
| `mci_gru/models/__init__.py` | `ResidualCrossSectionBlock` was importable only from the submodule while its sibling `SelfAttention` is exported. One line in `__all__`. |
| `docs/TEST_REGISTRY.md` | Required by this issue's own acceptance criteria, which the original owned-paths list contradicted. |

Nothing else moved. `configs/experiment/`, `docs/DEFAULT_EXPERIMENT_RECIPE.md`, and `paper_trade/` are untouched.

**Two review findings changed the implementation**, both worth recording:

1. **The checkpoint-compatibility guard was tautological.** It saved and reloaded a state dict from post-change code, so any regression in the default path would appear on both sides and the test would still pass. It now pins literals measured on `origin/main` at `125abda` before the change: 80 state-dict keys, 5,137 parameters, and the exact six-key `self_attention` set.
2. **`use_pre_norm` was scope creep.** It was unreachable from config and existed only to be tested. Removed; the LayerNorm is now unconditional and is pinned behaviourally instead, by asserting the attention's correction is unchanged when a stock's whole vector is shifted by a constant. That is the property the norm buys, and it fails without it.

**One finding not actioned.** The review noted that `ModelConfig.__post_init__` now carries a third near-identical enum-validation block and suggested collapsing all three. Declined as out of scope: it would touch validation for `output_activation` and `temporal_encoder`, which this issue does not own, and refactoring belongs to a separate change.
